### PR TITLE
Add ping service in basic host and libp2p constructor option (enabled by default)

### PR DESCRIPTION
**Pull request by vyzo**
Adds the ping service to basic_host, with a libp2p constructor option to control its instantiation; it's enabled by default.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -56,6 +56,9 @@ type Config struct {
 	NATManager  NATManagerC
 	Peerstore   pstore.Peerstore
 	Reporter    metrics.Reporter
+
+	PingCustom bool
+	Ping       bool
```
Comment (raulk) [inline config/config.go]: Not sure I'm liking the proliferation of the `...Custom` fields. If we make the actual flags pointers `*bool`, a `nil` could indicate no user-provided value. WDYT? Same for Relay/RelayCustom?

File: p2p/protocol/ping/ping_test.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-package ping
+package ping_test
```
Comment (raulk) [inline p2p/protocol/ping/ping_test.go]: Why is this change needed?

Comment (vyzo) [reply p2p/protocol/ping/ping_test.go]: we get an import cycle without it, because the test uses basic hosts.

File: config/config.go (hunk 1 above)
Comment (vyzo) [reply config/config.go]: I tried with a ternary enum, but it was disgusting to work with :)

**Comment by vyzo** [comment]
IPFS already enables ping, but they construct the PingService explicitly; this change obviates the need for that construction.
It is also very useful for autorealy, as we'd like to eventually choose relays based on latency.
I don't think there is a such a great concern for ping DoS right now, but let's open an issue to discuss the implications.

File: config/config.go (hunk 1 above)
Comment (raulk) [reply config/config.go]: How about modelling the ternary state like this? https://github.com/libp2p/go-libp2p/commit/4758a9552e20447bf34e7543d7607e7a5fb9ce76

Comment (vyzo) [reply config/config.go]: ugh, that's ugly too. I don't mind the Custom thingies that much, because the struct is only used for configuring the host and nothing else.

File: p2p/protocol/ping/ping_test.go (hunk 2 above)
Comment (lanzafame) [reply p2p/protocol/ping/ping_test.go]: I will not that this is a good practice for testing a package from the perspective of a 'user' of a package as it only gives you access to the exported functionality of the `ping` package. Though it can make testing certain things more difficult especially in the domain of ipfs and libp2p.

Comment (lanzafame) [reply p2p/protocol/ping/ping_test.go]: I will note, that this is a good practice for testing a package from the perspective of a 'user' of a package as it only gives you access to the exported functionality of the package. Though it can make testing certain things more difficult especially in the domain of ipfs and libp2p.

File: config/config.go (hunk 1 above)
Comment (marten-seemann) [reply config/config.go]: Why not `DisablePing`?
